server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -439,61 +404,319 @@ public static boolean tryFastFilterAggregation(
+        final int size
+    ) throws IOException {
+        // ranges are connected and in ascending order
+        Iterator<long[]> rangeIter = Arrays.stream(ranges).iterator();
+        long[] activeRange = rangeIter.next();
+
+        // make sure the first range at least crosses the min value of the tree
+        boolean noRangeMatches = false;
+        if (activeRange[0] > NumericUtils.sortableBytesToLong(tree.getMaxPackedValue(), 0)) {
+            noRangeMatches = true;
+        } else {
+            while (activeRange[1] < NumericUtils.sortableBytesToLong(tree.getMinPackedValue(), 0)) {
+                if (rangeIter.hasNext()) {
+                    activeRange = rangeIter.next();
+                } else {
+                    noRangeMatches = true;
                     break;
                 }
             }
         }
+        DebugInfo debugInfo = new DebugInfo();
+        if (noRangeMatches) {
+            logger.debug("No ranges match the query, skip the fast filter optimization");
+            return debugInfo;
+        }
 
-        logger.debug("Fast filter optimization applied to shard {} segment {}", fastFilterContext.context.indexShard().shardId(), ctx.ord);
-        return true;
+        RangeCollectorForPointTree collector = new RangeCollectorForPointTree(incrementDocCount, fieldType, rangeIter, size, activeRange);
+
+        final ArrayUtil.ByteArrayComparator comparator = ArrayUtil.getUnsignedComparator(8);
+        PointValues.IntersectVisitor visitor = getIntersectVisitor(collector, comparator);
+        try {
+            intersectWithRanges(visitor, tree, collector, debugInfo);
+        } catch (CollectionTerminatedException e) {
+            logger.debug("Early terminate since no more range to collect");
+        }
+        collector.finalizePreviousRange();
+
+        return debugInfo;
     }
 
-    private static boolean segmentMatchAll(SearchContext ctx, LeafReaderContext leafCtx) throws IOException {
-        Weight weight = ctx.searcher().createWeight(ctx.query(), ScoreMode.COMPLETE_NO_SCORES, 1f);
-        return weight != null && weight.count(leafCtx) == leafCtx.reader().numDocs();
+    private static void intersectWithRanges(
+        PointValues.IntersectVisitor visitor,
+        PointValues.PointTree pointTree,
+        RangeCollectorForPointTree collector,
+        DebugInfo debug
+    ) throws IOException {
+        PointValues.Relation r = visitor.compare(pointTree.getMinPackedValue(), pointTree.getMaxPackedValue());
+
+        switch (r) {
+            case CELL_INSIDE_QUERY:
+                collector.countNode((int) pointTree.size());
+                debug.visitInner();
+                break;
+            case CELL_CROSSES_QUERY:
+                if (pointTree.moveToChild()) {
+                    intersectWithRanges(visitor, pointTree, collector, debug);
+                    pointTree.moveToSibling();
+                    intersectWithRanges(visitor, pointTree, collector, debug);
```

Comment:
Just trying to make the code simple. I think the tree here is a simple binary tree. But let me know if I'm missing anything.

Reply:
As per my understanding, BKD tree is storage efficient version of KDB trees, which are nothing but k-dimensional b+ trees. Also, limiting the number of children to 2 probably bloats the number of internal nodes, but I might be wrong. The source of truth should be this one - https://users.cs.duke.edu/~pankaj/publications/papers/bkd-sstd.pdf

Reply:
It's even harder to confirm, since lucene may not implement it as the paper described. 
I will change it to the while loop.

Reply:
In practice, AFAIK, Lucene (currently) always uses a binary tree, but it's not guaranteed. Also, there are cases where the tree can become unbalanced, so it may be possible to only have 1 child.

So the `while` loop is safer both for number of children `< 2` as well as `> 2`.